main/contracts/erc20.ts

```diff
@@ -72,25 +75,19 @@ export default class Erc20Contract {
   }
 
   async getTokenData() {
-    try {
-      const calls = await Promise.all([
-        this.contract.decimals(),
-        this.contract.name(),
-        this.contract.symbol()
-      ])
+    const calls = await Promise.allSettled([
+      this.contract.decimals(),
+      this.contract.name(),
+      this.contract.symbol(),
+      this.contract.totalSupply() //totalSupply is mandatory on the ERC20 interface
+    ])
 
-      return {
-        decimals: calls[0],
-        name: calls[1],
-        symbol: calls[2]
-      }
-    } catch (e) {
-      log.error(`getTokenData error: ${e}`)
-      return {
-        decimals: 0,
-        name: '',
-        symbol: ''
-      }
+    const totalSupply = calls[3].status === 'fulfilled' ? calls[3].value.toString() : ''
+    return {
+      decimals: takeDefaultIfRejected(calls[0], 0),
+      name: takeDefaultIfRejected(calls[1], ''),
+      symbol: takeDefaultIfRejected(calls[2], ''),
+      totalSupply
```

Comment:
don't need a special function for this, you can just use catch. you can also use `Promise.all()` as well since none of these will reject if we're providing defaults:
```
const calls = await Promise.allSettled([
      this.contract.decimals().catch(() => 0),
      this.contract.name().catch(() => ''),
      this.contract.symbol().catch(() => ''),
      this.contract.totalSupply().then((supply: BigNumber) => supply.toString()).catch(() => '') // totalSupply is mandatory on the ERC20 interface
    ])

    return {
      decimals: calls[0],
      name: calls[1],
      symbol: calls[2],
      totalSupply: calls[3]
    }
```

Reply:
and if you want to get real fancy you can combine it all like this:
```
const calls = await Promise.all([
      ['decimals', this.contract.decimals().catch(() => 0)],
      ['name', this.contract.name().catch(() => '')],
      ['symbol', this.contract.symbol().catch(() => '')],
      ['totalSupply', this.contract.totalSupply().then((supply: BigNumber) => supply.toString()).catch(() => '')] // totalSupply is mandatory on the ERC20 interface
    ])

    return Object.fromEntries(calls)
```